File: backend/src/main/java/com/finance/tracker/entity/Goal.java
package com.finance.tracker.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.OffsetDateTime;

@Entity
@Getter
@Setter
@NoArgsConstructor
@Table(name = "goals")
public class Goal {
    public enum GoalStatus {
        IN_PROGRESS,
        COMPLETED,
        PAUSED
    }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private BigDecimal targetAmount;

    @Column(nullable = false)
    private BigDecimal currentAmount = BigDecimal.ZERO;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "linked_account_id")
    private Account linkedAccount;

    @Column
    private String icon;

    @Column
    private String color;

    @Column(nullable = false)
    private OffsetDateTime dueDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private GoalStatus status = GoalStatus.IN_PROGRESS;
}

File: frontend/src/types/goal.ts
export type GoalStatus = 'IN_PROGRESS' | 'COMPLETED' | 'PAUSED';

export type GoalResponse = {
  id: number;
  name: string;
  targetAmount: number;
  currentAmount: number;
  progressPercent: number;
  targetDate: string;
  linkedAccountId?: number;
  linkedAccountName?: string;
  icon?: string;
  color?: string;
  status: GoalStatus;
};

export type GoalRequest = {
  name: string;
  targetAmount: number;
  targetDate: string;
  linkedAccountId?: number;
  icon?: string;
  color?: string;
};

export type GoalContributionRequest = {
  amount: number;
  sourceAccountId?: number;
};

export type GoalWithdrawRequest = {
  amount: number;
  targetAccountId?: number;
};
